### PR TITLE
W3c trace context

### LINKED ISSUE #169: [W3C] Design and implement a generic API to create/accept distributed tracing payloads (CAT/DT/W3C)
Customers need a way to generate and accept W3C headers in the case where they don’t have access to the request/response or the framework doesn’t use HTTP.

We need to design and implement an API that includes two methods for propagating generic distributed trace header context. In particular the goal is to support W3C trace context headers (in addition to our existing support for New Relic CAT (Cross Application Tracing) and New Relic Distributed Tracing headers) but we want to design the APIs in a manner so that they can easily support new standards that may emerge. 

The API needs to adhere to the [DT agent spec](https://source.datanerd.us/agents/agent-specs/blob/master/distributed_tracing/Distributed-Tracing.md#api-names) and provide one method for generating the DT headers and one for DT accepting headers:

> The API **SHOULD** provide methods for accepting Distributed Trace payloads, and for inserting them into existing maps or HTTPHeader objects. If the methods are provided, they **MUST** use `exclude_newrelic_header` to decide whether or not to add the newrelic header.

>The API that is chosen will depend on the language conventions, but the names (barring camel-case vs. snake-case language differences) and signatures (different argument ordering for optional arguments are ok) should be similar to:

>`AcceptDistributedTraceHeaders(TransportType, Headers)`
>`InsertDistributedTraceHeaders(Headers)`

>Note that supportability metrics do not change and still should follow the naming guidelines as described [here](https://source.datanerd.us/agents/agent-specs/blob/master/distributed_tracing/New-Relic-Payload.md#supportability-metrics).

>The `InsertDistributedTraceHeaders` method **MUST** add HTTP-safe content into the headers - i.e. if the agent in the past has had options for `Text` or `HTTPSafe` payloads (where the `Text` version might be more compact), in this method the agent must insert HTTP-safe formatted information.

---

A Java specific implementation might look something like this:

`public insertDistributedTraceHeaders(Map<Object, Object>);`

Insert would inject headers into the passed in map so the user could propagate them however they want.

`public boolean acceptDistributedTraceHeaders(TransportType, Map<String, String>);`

Accept will take a [transport type](https://github.com/newrelic/newrelic-java-agent/blob/main/agent-bridge/src/main/java/com/newrelic/agent/bridge/TransportType.java) and map and parse out the correct payloads from the map and link the distributed trace and report the transport type with the transaction (i believe, it might also be the span). This leaves it to the user on how to get the payloads.

By keeping it general (using a `Map`) we can change the underlying behavior to support both W3C and our custom New Relic distributed trace payloads. Currently we send both payload types anyway so we should generate both payload types when calling insert.

Maintaining support for existing DT types with the new APIs should allows us to deprecate the old APIs.

---

Success criteria:
* Customers can create and accept W3C payloads over HTTP (in addition to existing supported payload types)
* Customers can create and accept W3C payloads over non-HTTP (in addition to existing supported payload types)
* Code is well tested. Tests may look similar to these [existing DT functional tests](https://github.com/newrelic/newrelic-java-agent/blob/main/functional_test/src/test/java/test/newrelic/test/agent/DistributedTracingTest.java).

### LINKED ISSUE #170: [W3C] Deprecate old existing DistributedTracePayload APIs
This work is dependent on #169 providing a general use API for creating/accepting distributed tracing payloads and rendering these older APIs redundant.

Once the new API is in place we should deprecate the [old create payload API](https://github.com/newrelic/newrelic-java-agent/blob/main/newrelic-api/src/main/java/com/newrelic/api/agent/Transaction.java#L301) as well as the [old accept payload API](https://github.com/newrelic/newrelic-java-agent/blob/main/newrelic-api/src/main/java/com/newrelic/api/agent/Transaction.java#L311-L321). Side note, [this one](https://github.com/newrelic/newrelic-java-agent/blob/main/newrelic-api/src/main/java/com/newrelic/api/agent/Transaction.java#L321) should be deprecated regardless because it’s not possible to use this unless you’re accepting a payload within the same service you created it (there’s no way currently a customer can turn a text payload into a `DistributedTracePayload` object).

---

Success criteria:
* These old APIs are marked as deprecated and the Javadoc comments point to the new APIs that replace them.

### PR DESCRIPTION
**Pull request by jack-berg**
A different take on the W3C trace context APIs using a newly introduced `Headers` interface. This is a unifying interface for `InboundHeaders` and `OutboundHeaders`, and makes up for a deficiency in `InboundHeader` that has necessitated `ExtendedInboundHeaders`. 

If we like this approach, we could deprecate `InboundHeaders` and `OutboundHeaders` and slowly transition all usages to `Headers`. Because `Headers` extends both `InboundHeaders` and `OutboundHeaders`, this transition could be done in small steps. 

I've also included a default implementation of the new headers interface called `ConcurrentHashMapHeaders`, which is thread safe and uses a `ConcurrentHashMap` as the backing data structure for the implementation. It seems necessary to provide users of the new APIs a reasonable default implementation for ease of use. 

Closes #169.
Closes #170.

### CODE REVIEW

**Comment by jack-berg** [comment]
I want to call out that my introduction of `ConcurrentHashMapHeaders` into the `newrelic-api` module seems to break convention a bit, since I'm not sure I see any other instantiatable objects defined in that module.

What's our stance on including classes in the `newrelic-api` module?

I can always move the class into some internal only module (it will be a useful tool for migrating away from `InboundHeaders` and `OutboundHeaders` internally), but I do think having it in published will ease the consumption of the APIs for our customers.

**Comment by jack-berg** [comment]
I've added an additional PR to resurrect an updated version of a test class for the new distributed trace APIs. 

I've also deprecated the old APIs.

**Review by jasonjkeller — APPROVED** [review]
Overall I like this approach better than the previous one. I think it results in a cleaner API and sets us up to deprecate/remove cruft going forward.

**Comment by jasonjkeller** [comment]
> I want to call out that my introduction of `ConcurrentHashMapHeaders` into the `newrelic-api` module seems to break convention a bit, since I'm not sure I see any other instantiatable objects defined in that module.
> 
> What's our stance on including classes in the `newrelic-api` module?
> 
> I can always move the class into some internal only module (it will be a useful tool for migrating away from `InboundHeaders` and `OutboundHeaders` internally), but I do think having it in published will ease the consumption of the APIs for our customers.

@jack-berg I think that our approach has generally been that if any critical problem occurs a customer should always be able to quickly remove the agent jar and all API calls in their code will simply default to a NoOp version. Because of this approach we can make the guarantee that there is zero impact on the customers environment when the agent jar is removed but API calls remain in their code.

While I'm having a hard time thinking of anything that might actually go wrong with the `ConcurrentHashMapHeaders` implementation included as is it might be best to take precautions and try our best to maintain such a guarantee. If we can make that guarantee with it in place (and I think we probably can) then I could be convinced to leave the implementation in place.

File: newrelic-agent/src/main/java/com/newrelic/agent/Transaction.java [hunk 1]
```diff
@@ -290,14 +290,6 @@ public DistributedTracePayloadImpl createDistributedTracePayload(String spanId)
         return payload;
     }
 
-    void publicApiAcceptDistributedTracePayload(String payload) {
-        acceptDistributedTracePayload(payload);
-    }
-
-    void publicApiAcceptDistributedTracePayload(DistributedTracePayload payload) {
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply newrelic-agent/src/main/java/com/newrelic/agent/Transaction.java:297]: You're seeing it correctly. I didn't understand either. Went back to the PR where the code was added and didn't find any context that indicated intent.

**Comment by jack-berg** [comment]
@jasonjkeller that seems like a good guarantee to make. In this case, I _believe_ the addition of `ConcurrentHashMapHeaders` should not jeopardize that. For both the insert and accept call, the client is responsible for providing a `Headers` instance. If the agent jar isn't on the classpath, I believe the no op implementation of `Transaction` will just receive the `Headers` instance and not do anything with them.

**Comment by jasonjkeller** [comment]
@tspring @XiXiaPdx Based on the discussion above, can either of you think of any reasons to not include the `ConcurrentHashMapHeaders` implementation in the public API? If not, I say we call this PR good and merge it.

Here's a practical example of the new APIs being used to propagate DT headers via Kafka record headers:
https://github.com/jasonjkeller/kafka-examples

**Review by tspring — APPROVED** [review]

**Pull request merged**